RV64 kern refactor

furquan-rivos opened the pull request.

furquan-rivos committed:
sys/kern: Reorganize the hierarchy for arch/

This change adds submodules to allow better organization under arch
for each architecture. This change is made to allow more files to be
added for a given architecture.

furquan-rivos committed:
sys/kern/rv64: Move SavedState to separate module

This change moves SavedState and associated methods to a separate
submodule `saved_state.rs`. This is done to improve the code
organization so that it's easier to add support for S-mode later.

furquan-rivos committed:
sys/kern/rv64: Drop the setting of CLOCK_FREQ_KHZ

CLOCK_FREQ_KHZ is already set by `start_kernel` so there is no need to
set it again in rv64 code.

furquan-rivos committed:
sys/kern/rv64: Move clock frequency handling to clock_freq.rs

This change moves clock frequeny handling function and saved variable
to separate submodule clock_freq.rs.

furquan-rivos committed:
sys/kern/rv64: Move macros to separate submodule

This change moves uassert macro to its own separate module to make
it easy to share with other submodules in rv64.

furquan-rivos committed:
sys/kern/rv64: Store current task pointer in mscratch

This change drops the global variable CURRENT_TASK_PTR and instead
stores the current task pointer in mscratch register. This drops
the usage of the variable in different functions and hence makes it
easier to split the files into separate submodules without having to
explicitly share this variable. This is in preparation to refactor
rv64 support into separate submodules in the future.

TEST=Verified that tests still pass for rv64.

Signed-off-by: Furquan Shaikh <furquan@rivosinc.com>

furquan-rivos committed:
sys/kern/rv64: Move task-related functions to task.rs

This change moves all task-related functions into its own submodule
`task.rs`.

furquan-rivos committed:
sys/kern/rv64: Move time related functions to mtimer.rs and ticks.rs

This change moves all functions dealing with time to two separate
submodules:
1. `mtimer.rs` - this deals with M-mode timer
2. `ticks.rs` - this deals with kernel ticks

This change is done to make it easier to support different modes for
kernel later on.

furquan-rivos committed:
sys/kern/rv64: Add helpers submodule

This change moves general helper code from `riscv64.rs` to a separate
submodule `helpers.rs`. Basically, these are the functions that didn't
really have a better place to go to.

furquan-rivos committed:
sys/kern/rv64: Ensure mscratch is set before using `get_current_task`

This change adds an assert if `get_current_task()` is called without
mscratch being set.

furquan-rivos committed:
sys/kern/rv64: Move pmp support to its own submodule

This change moves `apply_memory_protection()` using pmp to its own
submodule `pmp.rs`. This is required to allow different models of
memory protection in the future depending upon the mode in which
kernel operates.

furquan-rivos committed:
sys/kern/rv64: mscratch current task ptr updates

This change performs following updates based on review feedback:
1. Use csrrw to read mscratch into a1 and a0 into mscratch instead of
   separate csrr and csrw.
2. Add comments to indicate that mscratch temporarily doesn't point to
   current task while stashing the current task status in
   _start_trap.

TEST=Verified that tests still pass for rv64.

furquan-rivos committed:
Merge branch 'rivos/main-rv64' into dev/furquan/rv64-mscratch-task-ptr

furquan-rivos committed:
sys/kern/rv64: Rename riscv64.rs to trap.rs

This change renames riscv64.rs to trap.rs since the file just deals
with traps after all the refactoring.

TEST=Verified with all the changes in this series that rv64 tests
still pass.

furquan-rivos commented:
I created this branch for PR from `dev/furquan/rv64-mscratch-task-ptr` since I wanted to add changes on top of the mscratch changes. When I raise this PR, the original changes also show up here for review. I don't have a better way to mark those initial changes as ignore in this PR.

furquan-rivos committed:
sys/kern/rv64/macros: Add missing license

furquan-rivos commented on sys/kern/src/arch/rv64/macros.rs:
```diff
@@ -0,0 +1,7 @@
+macro_rules! uassert {
```
Done.

furquan-rivos commented on sys/kern/src/arch.rs:
```diff
@@ -21,16 +21,16 @@ cfg_if::cfg_if! {
+        pub mod arm;
+        pub use arm::*;
     } else if #[cfg(target_arch = "riscv32")] {
         #[macro_use]
```
At some point, I think we can put `rv32` and `rv64` under a single `rv` and create a hierarchy under that. But, I didn't do that now to make it easier for rv32 changes to be pushed upstream for review.

furquan-rivos commented on sys/kern/src/arch/rv64/mtimer.rs:
```diff
@@ -0,0 +1,56 @@
+
+pub fn reset_timer() {
+    unsafe {
+        core::ptr::write_volatile(MTIME as *mut u64, 0);
```
Raised a [bug](https://rivosinc.atlassian.net/browse/SW-605) to fix this.

furquan-rivos commented on sys/kern/src/arch/rv64/mtimer.rs:
```diff
@@ -0,0 +1,56 @@
+// On both RV32 and RV64 systems the mtime and mtimecmp memory-mapped registers
+// are 64-bits wide.
+//
+const MTIMECMP: u64 = 0x0200_4000;
```
Raised a [bug](http://go/ji/SW-606) to fix this hardcoding.

FawazTirmizi commented on sys/kern/src/arch/rv64/ticks.rs:
```diff
@@ -0,0 +1,20 @@
+#[used]
+static mut TICKS: u64 = 0;
+
+pub fn incr_ticks(incr: u64) -> Timestamp {
+    let ticks = unsafe { &mut TICKS };
+    *ticks += incr;
+    drop(ticks);
+    now()
+}
+
+/// Reads the tick counter.
+pub fn now() -> Timestamp {
+    Timestamp::from(unsafe { TICKS })
+}
```
Maybe these should be #[inline(always)]? I'd expect them to get inlined in most cases, but if it doesn't the call overhead may be expensive.

furquan-rivos replied:
I was leaving it up to the rust compiler to decide that. But, I can add that since we want to strongly suggest inlining.

furquan-rivos replied:
Pushed a follow up commit to fix this.

furquan-rivos committed:
sys/kern/rv64: Suggest strongly to inline tick functions

Tick functions get used very often and so we would like to suggest
strongly to the compiler to inline these functions.

furquan-rivos committed:
Merge branch 'rivos/main-rv64' into dev/furquan/rv64-kern-refactor
